target/linux/ipq40xx/base-files/etc/board.d/02_network

```diff
@@ -235,6 +235,12 @@ ipq40xx_setup_macs()
 	pakedge,wr-1)
 		wan_mac=$(macaddr_add $(get_mac_label) 1)
 		;;
+	tel,x1pro)
+		label_mac=$(mtd_get_mac_binary "0:ART" 0x1006)
+		wan_mac=$(macaddr_add "$label_mac" 1)
+		lan_mac=$(macaddr_add "$label_mac" 2)
+		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3 lan4" "wan"
+		;;
```

Comment:
Please use nvmem cell and declare this in dts.

Reply:
Okay I will try. I don't know what I'm doing though, any documentation you can point me to?

Reply:
There are plenty of examples in the target, look at this for start:
https://github.com/openwrt/openwrt/commit/4c5cb5880573b7e39a205f220f4f569db542973d

Should be easy since pre-cal is already being fetched from ART

Reply:
Im not sure how to access 0x1006 inside the ART partition. This is what I have tried, but it's not compiling. I put the whole partition for context

```c
	art: partition@170000 {
		label = "0:ART";
		reg = <0x170000 0x10000>;
		read-only;

		nvmem-layout {
			compatible = "fixed-layout";
			#address-cells = <1>;
			#size-cells = <1>;

			precal_art_1000: precal@1000 {
				reg = <0x1000 0x2f20>;
			};

			precal_art_5000: precal@5000 {
				reg = <0x5000 0x2f20>;
			};

			macaddr_base: precal@1006 { // add this?
				compatible = "mac-base";
				reg = <0x1006 0x6>;
				#nvmem-cell-cells = <1>;
			};
		};
	};
```

```c
&gmac {
	status = "okay";

	nvmem-cell-names = "mac-address"; 
	nvmem-cells = <&macaddr_base 2>; // access like this? (with the increment by 2 for this one)
};
```

Reply:
Tried this as well. I dont think it matters for the wifi because they are getting their MAC address directly from the ART config. But the MAC addresses for the LAN and WAN are not being set.
```c
	art: partition@170000 {
		label = "0:ART";
		reg = <0x170000 0x10000>;
		read-only;

		nvmem-layout {
			compatible = "fixed-layout";
			#address-cells = <1>;
			#size-cells = <1>;

			macaddr_art_1006: macaddr@1006 {
				compatible = "mac-base";
				reg = <0x1006 0x6>;
				#nvmem-cell-cells = <1>;
			};

			precal_art_1000: precal@1000 {
				reg = <0x1000 0x2f20>;
			};

			precal_art_5000: precal@5000 {
				reg = <0x5000 0x2f20>;
			};

			macaddr_art_5006: mac-address@5006 {
				compatible = "mac-base";
				reg = <0x5006 0x6>;
				#nvmem-cell-cells = <1>;
			};
		};
	};
```

```c
&wifi0 {
	status = "okay";
	nvmem-cell-names = "pre-calibration, mac-address";
	nvmem-cells = <&precal_art_1000>, <&macaddr_art_1006>;
};

&wifi1 {
	status = "okay";
	nvmem-cell-names = "pre-calibration, mac-address";
	nvmem-cells = <&precal_art_5000>, <&macaddr_art_5006>;
};

&gmac {
	status = "okay";

	nvmem-cell-names = "mac-address";
	nvmem-cells = <&macaddr_art_1006 2>;
};

&swport5 {
	status = "okay";

	label = "wan";
	nvmem-cell-names = "mac-address";
	nvmem-cells = <&macaddr_art_1006 1>;
};
```

Reply:
MAC-s are supposed to be at `0x0` and most likely also at `0x6` in ART.

Look at:
https://github.com/openwrt/openwrt/blob/main/target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4029-gl-b1300.dts#L207

Reply:
`0x0`? I'm not sure. The MAC address for these devices starts from `0x1006` and `0x5006`. In the hexdump output I put it in <>. I thought what I was doing here is tell it to read 6 bytes starting from offset 0x1006, am I wrong?

```c
	macaddr_art_1006: macaddr@1006 {
		compatible = "mac-base";
		reg = <0x1006 0x6>;
		#nvmem-cell-cells = <1>;
	};
```

```shell
root@OpenWrt:/# hexdump -C /dev/mtd7
00000000  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00001000  20 2f af 8e 01 01 <70 b3  d5 26 06 99> 00 00 20 00  | /....p..&.... .|
00001010  2a a0 00 01 00 60 00 00  0a 07 10 00 00 00 00 00  |*....`..........|
```

Reply:
Ok, this makes more sense, it seems that vendor did not populate them properly as 0x0, 0x6 etc in ART are meant to be populated with ethernet MAC-s.

Hm, what you did should be correct.

Reply:
Lol, thanks; I am not too sure where to go from here. If it is properly setting the MAC then I should at least get the same wrong MAC address each time, but it randomly generates one each boot. Do I need to do anything else in the board setup scripts? Should I revert to just using the board setup script?

Reply:
I'd still like to resolve this if possible. 

```shell
root@Openwrt:~# hexdump -C /dev/mtd7 | grep "70 b3  d5"
00001000  20 2f ae 97 01 01 <70 b3  d5 26 03 75> 00 00 20 00  | /....p..&.u.. .|
00005000  20 2f 27 79 01 01 <70 b3  d5 26 03 74> 00 00 20 00  | /'y..p..&.t.. .|
```

Reply:
Does it print an error or just randomly uses MAC-s?

Reply:
It just uses randomly generated MAC addresses